pkg/sql/types/types.go

```diff
@@ -1186,22 +1178,31 @@ func (t *T) TupleLabels() []string {
 	return t.InternalType.TupleLabels
 }
 
-// StableTypeID returns the stable ID of the TypeDescriptor that backs this
-// type. This function only returns non-zero data for user defined types.
-func (t *T) StableTypeID() uint32 {
-	return t.InternalType.UDTMetadata.StableTypeID
+// UserDefinedArrayOID returns the OID of the array type that corresponds to
+// this user defined type. This function only can only be called on user
+// defined types and returns non-zero data only for user defined types that
+// aren't arrays.
+func (t *T) UserDefinedArrayOID() oid.Oid {
+	return t.InternalType.UDTMetadata.ArrayTypeOID
+}
+
+// SetOID sets the OID of the input type. It mutates the input type and should
```

Comment:
Instead do something like this:
```
func RemapOIDs(typ *T, newOid, newUserDefinedArrayOid oid.Oid) {
  typ.InternalType.Oid = newOid
  type.InternalType.UDTMetadata.ArrayTypeOID = newUserDefinedArrayOid
}
```
The reasons why this is better from an API design point of view:
1. The `Set` methods, which should almost never be used, won't clutter the `types.T` interface, and won't pop up as a completion suggestion on a `types.T` instance.
2. The purpose of setting these OIDs is more clearly stated in the name of the function. `Set` invites future developers to abuse the interface and violate immutability. `Remap` at least will give some clue that maybe the only time you should call this is if you're remapping OIDs, which presumably would be rare.

These details may seem inconsequential, but it's the details that separate a great API from a lousy one. See https://nypost.com/2016/03/12/infamous-botched-jesus-painting-now-a-major-tourist-attraction/ for how the little details can make a big difference. : )

Reply:
I tried something like this, but it just feels dirtier due to how we actually have to do the remapping. We only know the array type OID to remap to if the type is actually an array type, so we'd have to pass an argument that is ignored in some cases, or just split the function into two cases like it already is. I understand where you're coming from here, and it makes sense. One thing that I've done now is prefaced the functions with `Unsafe` and changed the names to `Remap` so that it's more clear the intended usecase.

Reply:
I think it's OK if the `newUserDefinedArrayOid` is just 0. That's what it'll be if you call `t.UserDefinedArrayOID`, right? In fact, you even have a comment implying the OID is zero for non-arrays: `returns non-zero data only for user defined types that aren't arrays`. We can do a bit of verification in the `RemapOIDs` function to validate these assumptions. I really don't like methods on the `types.T` interface for this, because I don't want to advertise this functionality. We want a small, unobtrusive back-door that's not particularly easy to find, and that does not advertise its presence.